lgsm/functions/workshop_core.sh

```diff
@@ -0,0 +1,318 @@
+			# Need tp purchase game.
+			elif [ -n "$(grep "No subscription" "${steamcmdlog}" | tail -1)" ]; then
+				fn_print_failure_nl "${commandaction} ${selfname}: ${remotelocation}: Steam account does not have a license for the required game"
+				fn_script_log_fatal "${commandaction} ${selfname}: ${remotelocation}: Steam account does not have a license for the required game"
+				core_exit.sh
+			# Two-factor authentication failure
+			elif [ -n "$(grep "Two-factor code mismatch" "${steamcmdlog}" | tail -1)" ]; then
+				fn_print_failure_nl "${commandaction} ${selfname}: ${remotelocation}: Two-factor authentication failure"
+				fn_script_log_fatal "${commandaction} ${selfname}: ${remotelocation}: Two-factor authentication failure"
+				core_exit.sh
+			# Incorrect Branch password
+			elif [ -n "$(grep "Password check for AppId" "${steamcmdlog}" | tail -1)" ]; then
+				fn_print_failure_nl "${commandaction} ${selfname}: ${remotelocation}: betapassword is incorrect"
+				fn_script_log_fatal "${commandaction} ${selfname}: ${remotelocation}: betapassword is incorrect"
+				core_exit.sh
+			# Update did not finish.
+			elif [ -n "$(grep "0x402" "${steamcmdlog}" | tail -1)" ] || [ -n "$(grep "0x602" "${steamcmdlog}" | tail -1)" ]; then
+				fn_print_error2_nl "${commandaction} ${selfname}: ${remotelocation}: Update required but not completed - check network"
+				fn_script_log_error "${commandaction} ${selfname}: ${remotelocation}: Update required but not completed - check network"
+			else
+				fn_print_error2_nl "${commandaction} ${selfname}: ${remotelocation}: Unknown error occured"
+				echo -en "Please provide content log to LinuxGSM developers https://linuxgsm.com/steamcmd-error"
+				fn_script_log_error "${commandaction} ${selfname}: ${remotelocation}: Unknown error occured"
+			fi
+		elif [ "${exitcode}" != "0" ]; then
+			fn_print_error2_nl "${commandaction} ${selfname}: ${remotelocation}: Exit code: ${exitcode}"
+			fn_script_log_error "${commandaction} ${selfname}: ${remotelocation}: Exit code: ${exitcode}"
+		else
+			fn_print_complete_nl "${commandaction} ${selfname}: ${remotelocation}"
+			fn_script_log_pass "${commandaction} ${selfname}: ${remotelocation}"
+		fi
+
+		if [ "${counter}" -gt "10" ]; then
+			fn_print_failure_nl "${commandaction} ${selfname}: ${remotelocation}: Did not complete the download, too many retrys"
+			fn_script_log_fatal "${commandaction} ${selfname}: ${remotelocation}: Did not complete the download, too many retrys"
+			core_exit.sh
+		fi
+	done
+}
+
+fn_workshop_get_list() {
+	workshoplist=($(echo "${workshopmods}" | tr ";" "\n"))
+}
+
+fn_workshop_get_latest_mod_version() {
+	local modid="$1"
+	local serverresp="$(curl -s -d "itemcount=1&publishedfileids[0]=${modid}" "http://api.steampowered.com/ISteamRemoteStorage/GetPublishedFileDetails/v1")"
+	local remupd=
+	if [[ "${serverresp}" =~ \"hcontent_file\":[[:space:]]*([^,]*) ]]; then
+       remupd="${BASH_REMATCH[1]}"
+    fi
+    echo "${remupd}" | tr -d '"'
```

Comment:
Instead of loading up tr(1) and using a subshell (via pipe) for this, it's far more efficient to use BASH's parameter expansion's pattern substitution, as it's already in the language and it's a very small amount of data.

```bash
echo "${remupd//\"/}"
```

Two initial '/' indicate a greedy match, while one is non-greedy.

This feature is available from at least BASH 3.0, including the option to omit the final '/' if no replacement string is provided.